codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customize/ServerCodegenDecorator.kt

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.server.smithy.customize
+
+import software.amazon.smithy.build.PluginContext
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.rust.codegen.core.smithy.RustCrate
+import software.amazon.smithy.rust.codegen.core.smithy.generators.LibRsCustomization
+import software.amazon.smithy.rust.codegen.core.smithy.generators.ManifestCustomizations
+import software.amazon.smithy.rust.codegen.core.smithy.protocols.ProtocolMap
+import software.amazon.smithy.rust.codegen.core.util.deepMergeWith
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
+import software.amazon.smithy.rust.codegen.server.smithy.generators.protocol.ServerProtocolGenerator
+import java.util.ServiceLoader
+import java.util.logging.Logger
+
+typealias ServerProtocolMap = ProtocolMap<ServerProtocolGenerator, ServerCodegenContext>
+
+/**
+ * [ServerCodegenDecorator] allows downstream users to customize code generation.
+ *
+ * For example, AWS-specific code generation generates customizations required to support
+ * AWS services. A different downstream customer may wish to add a different set of derive
+ * attributes to the generated classes.
```

Comment:
Nit: the AWS example doesn't really fit for the server:

```suggestion
 * For example, a downstream customer may wish to add a different set of derive
 * attributes to the generated classes.
```